### PR TITLE
[MRG] Speedup

### PR DESCRIPTION
**Pull request by vene**
Switch everything to typed memoryviews and inline direct solver update.

```
20 newsgroups
=============
X_train.shape = (11314, 130107)
X_train.format = csc
X_train.dtype = float64
X_train density = 0.001214353154362896
y_train (11314,)
X_test (7532, 130107)
X_test.format = csc
X_test.dtype = float64
y_test (7532,)

Classifier Training
===================
/home/vlad/code/polylearn/polylearn/factorization_machine.py:118: UserWarning: Objective did not converge. Increase max_iter.
  warnings.warn("Objective did not converge. Increase max_iter.")
Training fm-2 ... done
Training fm-3 ... done
/home/vlad/code/polylearn/polylearn/polynomial_network.py:99: UserWarning: Objective did not converge. Increase max_iter.
  warnings.warn("Objective did not converge. Increase max_iter.")
Training polynet-2 ... done
Training polynet-3 ... done
Classification performance:
===========================

Classifier            train       test         f1   accuracy
------------------------------------------------------------
fm-3               36.3608s    0.3642s     0.4009     0.9663
fm-2               31.7401s    0.1225s     0.6126     0.9740
polynet-3           6.6819s    0.0685s     0.6992     0.9796
polynet-2           5.9581s    0.0693s     0.7310     0.9807

>>> on branch speedup

Classifier            train       test         f1   accuracy
------------------------------------------------------------
fm-3               11.7805s    0.3608s     0.4009     0.9663
fm-2                7.1945s    0.1218s     0.6126     0.9740
polynet-3           6.5308s    0.0676s     0.6992     0.9796
polynet-2           5.9655s    0.0675s     0.7310     0.9807

```

### CODE REVIEW

**Comment by vene** [comment]
@fabianp could you give me a hand to figure out why I'm getting [these win32/py2 failures](https://ci.appveyor.com/project/vene/polylearn/build/1.0.29)? It looks like the `cd_direct_fast.pyx` solver has numerical issues in this configuration, but I can't pinpoint the cause.

**Comment by fabianp** [comment]
Hey. Went through the file cd_direct_fast.pyx but I don't know what can be causing such failures. Any chance you could make a minimal example for such failures?

**Comment by mblondel** [comment]
Since the win32 failures are independent of this PR, I think this PR should be merged, since it is a big improvement. I would also add a benchmark/ folder to help track the improvements over time.

Regarding the failures, could it be that the random kit behave differently on win32?

**Comment by fabianp** [comment]
If that is the case, adding l2 regularization and/or more iterations should
solve the problem

**Comment by mblondel** [comment]
Remember that FMs and PNs are non-convex :)

**Comment by mblondel** [comment]
@ogrisel From past experience with scikit-learn, any idea what the win32 failures could come from?

By the way, the problem is only on Python 2.7. Python 3.5 is fine.

**Comment by fabianp** [comment]
D'oh!

File: polylearn/cd_direct_fast.pyx [hunk 1]
```diff
@@ -18,20 +18,20 @@ from .cd_linear_fast cimport _cd_linear_epoch
 
 cdef void _precompute(ColumnDataset X,
                       double[:, :, ::1] P,
-                      int order,
+                      Py_ssize_t order,
                       double* out,
-                      int s,
-                      int degree):
```
Comment (ogrisel) [inline polylearn/cd_direct_fast.pyx]: It's weird to use `Py_ssize_t` for a mathematical variable. `Py_ssize_t` should only be used to index arrays.

**Comment by ogrisel** [comment]
`Py_ssize_t` has a different meaning signed 32 bit int on a 32 bit platform and signed 64 bit int on a 64 bit platform. It should only be used to index datastructures (e.g. arrays or Python lists). I do not understand why the problem does not exist with Python 3 though.

**Comment by vene** [comment]
That makes sense. Thanks, @ogrisel! However the problem still existed when I was using int everywhere. I only replaced it with py_ssize_t in an attempt to fix it, and it seems I got a bit overzealous and replaced too many vars.

I'll change it to use py_ssize_t just where it should.

On July 25, 2016 10:23:48 AM EDT, Olivier Grisel <notifications@github.com> wrote:
>`Py_ssize_t` has a different meaning signed 32 bit int on a 32 bit
>platform and signed 64 bit int on a 64 bit platform. It should only be
>used to index datastructures (e.g. arrays or Python lists). I do not
>understand why the problem does not exist with Python 3 though.
>
>---
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/scikit-learn-contrib/polylearn/pull/1#issuecomment-234967790

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by vene** [comment]
@mblondel I think it's not a rng issue. The `y` generated in `tests/test_factorization_machine.py:check_fit` for both `degree=2, 3` is the same in AppVeyor as on my 64bit ubuntu machine. I haven't directly checked the initial P_ but I doubt it would be different.

I also fixed the types as suggested by @ogrisel with no difference.

So it's probably something numerical.  It's strange that no errors occur with the lifted solver though.

Maybe I should try a 32bit py2 env in linux and see if I can reproduce it.

File: polylearn/cd_direct_fast.pyx [hunk 2]
```diff
@@ -71,14 +75,14 @@ cdef inline double _update(int* indices,
         if degree == 2:
             grad_y = d1[i] - p_js * data[ii]
         elif degree == 3:
-            grad_y = 0.5 * (d1[i] ** 2 - d2[i])
```
Comment (mblondel) [inline polylearn/cd_direct_fast.pyx]: I am surprised that Cython doesn't optimize this directly. Did it make a significant difference?

**Comment by mblondel** [comment]
Is there a way to reduce the verbosity of coveralls? I receive one email per commit, which makes it a bit difficult to follow real comments.

**Comment by vene** [comment]
Ok, I think I made coveralls be quiet.

File: polylearn/cd_direct_fast.pyx (hunk 2 above)
Comment (vene) [inline polylearn/cd_direct_fast.pyx]: I was looking at the generated C++ and found that `a ** 2` is compiled into `pow(a, 2)`.  However, it makes no difference in terms of speed. I guess the compiler inlines `pow` calls efficiently.

Comment (mblondel) [inline polylearn/cd_direct_fast.pyx]: If there's no measurable speed up (which confirms my experience), I would favor the readable version.

Comment (vene) [inline polylearn/cd_direct_fast.pyx]: I agree, I'll revert.

Meanwhile a different win64+py2 failure appeared out of nowhere (?), I'll
try actually investigating on a windows machine.

**Comment by vene** [comment]
I fixed the windows bug! It was my fault, of course. I forgot to initialize a `double` scalar to 0, and it turns out this doesn't behave well on MSVC (oddly, only for py2, not for py3: must be because of the different compiler version?!)

Appveyor is green now for both win32 and win64. (I canceled a duplicate build, triggered by the fact that this branch is on scikit-learn-contrib rather than on my fork, feel free to ignore that.)

I'll remove the ugly `_sq` thing, update the benchmarks, then ping again for a final review before merging.

**Comment by vene** [comment]
This is now ready to merge.

**Comment by mblondel** [comment]
> I fixed the windows bug!

Hurray! Squash and merge?

**Comment by vene** [comment]
[Merged by rebase.](https://github.com/scikit-learn-contrib/polylearn/commit/195f4cd493564d18c7725acb40d9f341dda5bb0a) :beers:

**Pull request closed without merging**